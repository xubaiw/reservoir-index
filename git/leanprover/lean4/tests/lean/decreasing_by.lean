mutual
  inductive Even : Nat → Prop
    | base : Even 0
    | step : Odd n → Even (n+1)
  inductive Odd : Nat → Prop
    | step : Even n → Odd (n+1)
end
decreasing_by assumption

mutual
 def f (n : Nat) :=
   if n == 0 then 0 else f (n / 2) + 1
 decreasing_by assumption
end

def g' (n : Nat) :=
  match n with
  | 0 => 1
  | n+1 => g' n * 3
decreasing_by
  h => assumption

namespace Test
mutual
  def f : Nat → α → α → α
    | 0, a, b => a
    | n+1, a, b => g n a b |>.1

  def g : Nat → α → α → (α × α)
    | 0, a, b => (a, b)
    | n+1, a, b => (h n a b, a)

  def h : Nat → α → α → α
    | 0, a, b => b
    | n+1, a, b => f n a b
end
decreasing_by
  f => assumption
  g => assumption

end Test
