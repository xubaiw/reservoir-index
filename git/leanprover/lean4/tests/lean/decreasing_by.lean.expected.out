decreasing_by.lean:8:0-8:24: error: invalid 'decreasing_by' in mutually inductive datatype declaration
decreasing_by.lean:13:1-13:25: error: invalid 'decreasing_by' in 'mutual' block, it must be used after the 'end' keyword
decreasing_by.lean:21:2-21:3: error: function 'h' not found in current declaration
decreasing_by.lean:39:2-39:17: error: invalid termination hint element, 'Test.g' and 'Test.f' are in the same clique
